src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs

```diff
@@ -841,58 +841,121 @@ private static bool IsConditionalCastSafeToRemove(
 
-        // when we have    a ? (T)b : c
-        // 
-        // then we want the type of the written conditional to be T as well.  And we want the final converted
-        // type of `a ? b : c` to be the same as what `a ? (T)b : c` is converted to.
+        // Returns true if we have `x ? (T)y : z` and (T) can be removed because the 'y' type is hte same as the 'z' type, and
+        // both are converted to 'T' outside of the conditional.
+        bool IsConditionalCastSafeToRemoveDueToConversionToOtherBranch()
+        {
+            // Always keep a cast of 'default'.  This can end up taking on incorrect values if it uses the type of hte other branch
+            // (for example, between `(int?)default` vs `(int)default`.
+            if (castExpression.Expression.WalkDownParentheses().IsKind(SyntaxKind.DefaultLiteralExpression))
+                return false;
 
-        if (!originalConditionalTypeInfo.ConvertedType!.Equals(rewrittenConditionalTypeInfo.ConvertedType, SymbolEqualityComparer.IncludeNullability))
-            return false;
+            var otherSide = parent == originalConditionalExpression.WhenFalse ? originalConditionalExpression.WhenTrue : originalConditionalExpression.WhenFalse;
+            var otherSideType = originalSemanticModel.GetTypeInfo(otherSide, cancellationToken).Type;
+            var thisSideRewrittenType = rewrittenSemanticModel.GetTypeInfo(rewrittenExpression, cancellationToken).Type;
 
-        var castType = originalSemanticModel.GetTypeInfo(castExpression, cancellationToken).Type;
-        if (IsNullOrErrorType(castType))
-            return false;
+            if (otherSideType is null || thisSideRewrittenType is null)
+                return false;
 
-        if (rewrittenSemanticModel.GetOperation(rewrittenConditionalExpression, cancellationToken) is not IConditionalOperation rewrittenConditionalOperation)
-            return false;
+            // Check if 'y' has the same type as 'z'.
+            if (!otherSideType.Equals(thisSideRewrittenType))
+                return false;
 
-        if (castType.Equals(rewrittenConditionalOperation.Type, SymbolEqualityComparer.IncludeNullability))
-            return true;
+            // Now check that with the (T) cast removed, that the outer `x ? y : z` is still immediately converted to a
+            // 'T'. If so, we can remove this inner (T) cast.
 
-        if (rewrittenConditionalOperation.Parent is IConversionOperation conditionalParentConversion &&
-            conditionalParentConversion.GetConversion().IsImplicit &&
-            castType.Equals(conditionalParentConversion.Type, SymbolEqualityComparer.IncludeNullability))
-        {
-            return true;
-        }
+            // Due to https://github.com/dotnet/roslyn/issues/76799, we can't use a speculative model here.   So go and
+            // fork the actual model so we can get accurate information.  When the above bug is fixed we should be able
+            // to swap this code out:
```

Comment:
painful.  will hopefully be able to remove this once compiler fixes the linked issue.